app/src/main/java/com/cmput301f20t21/bookfriends/ui/add/AddEditViewModel.java

```diff
@@ -72,43 +63,31 @@ public void handleAddBook(
-            @Nullable Uri newUri, OnSuccessCallbackWithMessage<Book> successCallback, OnFailCallback failCallback
+            @Nullable Uri newUri, OnSuccessCallbackWithMessage<Book> successCallback, OnFailCallbackWithMessage<BOOK_ERROR> failCallback
     ) {
-            String bookId = oldBook.getId();
-            bookRepository.editBook(bookId, isbn, title, author, description).addOnCompleteListener(
-                    editBookTask -> {
-                        if (editBookTask.isSuccessful()) {
-                            Book book = new Book(bookId, isbn, title, author, description, oldBook.getOwner(), BOOK_STATUS.AVAILABLE, newUri);
-                            Uri oldUri = oldBook.getImageUri();
-                            if (newUri != null) {
-                                // image not changed
-                                if (oldUri != null && oldUri.equals(newUri)) {
-                                    successCallback.run(book);
-                                } else { // image is being added or replaced
-                                    String imageName = bookId + "cover";
-                                    // addImage will also replace if file with imageName already exist
-                                    bookRepository.addImage(imageName, newUri).addOnCompleteListener(
-                                            addImageTask -> {
-                                                if (addImageTask.isSuccessful()) {
-                                                    bookRepository.addImageNameToBook(bookId, imageName).addOnCompleteListener(
-                                                            addNameTask -> successCallback.run(book)
-                                                    );
-                                                } else {
-                                                    failCallback.run(BOOK_ERROR.FAIL_TO_ADD_IMAGE);
-                                                }
-                                            }
-                                    );
+        String bookId = oldBook.getId();
+        bookRepository.editBook(bookId, isbn, title, author, description).addOnCompleteListener(
+                editBookTask -> {
+                    if (editBookTask.isSuccessful()) {
+                        Book book = new Book(bookId, isbn, title, author, description, oldBook.getOwner(), BOOK_STATUS.AVAILABLE);
+                        String imageName = bookId + "cover";
+                        // addImage will also replace if file with imageName already exist
+                        bookRepository.addImage(imageName, newUri).addOnCompleteListener(
+                                addImageTask -> {
+                                    if (addImageTask.isSuccessful()) {
+                                        bookRepository.addImageNameToBook(bookId, imageName).addOnCompleteListener(
```

Comment:
no need to keep image name in book document?

Reply:
good catch